flake.nix

```
@@ -0,0 +1,111 @@
+                  # files. However, because all python package deps are propagated py poetry2nix into any
+                  # subsequent packages that depend on them, we get a clash when two versions of the same package occur. 
+                  # Below, we manually filter out the filelock package locked by pyk's poetry.lock file
+                  # and substitute in the version from mir's poetry.lock.
+                  # This is not ideal and may not be feasible if there are more/more complex clashes.
+                  # We could just let poetry2nix download pyk without importing it as a flake,
+                  # but then we would lose the ability to do kup install kmir --override pyk <version>
+                  # and we would instead have to modify pyproject.toml to point to the version of pyk we want, then
+                  # call kup install kmir --version . to use the modified local checkout.
+                  propagatedBuildInputs = prev.lib.filter
+                    (x: !(prev.lib.strings.hasInfix "filelock" x.name))
+                    old.propagatedBuildInputs ++ [ finalPython.filelock ];
+                });
+              });
+            groups = [ ];
+            # We remove `dev` from `checkGroups`, so that poetry2nix does not try to resolve dev dependencies.
+            checkGroups = [ ];
+          };
+        })
+      (final: prev: {
+          mir-semantics = prev.stdenv.mkDerivation {
+            pname = "mir-semantics";
+            version = self.rev or "dirty";
+            buildInputs = with prev; [
+              k-framework.packages.${prev.system}.k
+              kmir
+            ];
+            nativeBuildInputs = [ prev.makeWrapper ]; 
+
+            src = ./kmir;
+            
+            # kmir init $(kbuild which llvm) is to populate $out/lib/llvm
+            # with parser_Mir_MIR-PARSER-SYNTAX from gen_glr_parser 
+            # before nix directory becomes read only
+            buildPhase = ''
+              export KBUILD_DIR=".kbuild"
+              make build-backends POETRY_RUN=
+              kmir init $(kbuild which llvm)
+            '';
+
+            installPhase = ''
+              mkdir -p $out/lib/
+              cp -r $(kbuild which llvm) $out/lib/
+              cp -r $(kbuild which haskell) $out/lib/ 
+
+              mkdir -p $out/bin/
+            '';
+          };
+        })
+        # Now mir-semantics is built, wrap kmir with LLVM and HASKELL defs
+        (final: prev: {
+          kmir = prev.kmir.overrideAttrs (oldAttrs: {
+            buildInputs = oldAttrs.buildInputs or [] ++ [ prev.python310 prev.makeWrapper ];
+            postInstall = oldAttrs.postInstall or "" + ''
+              wrapProgram $out/bin/kmir --set KMIR_LLVM_DIR ${(toString prev.mir-semantics) + "/lib/llvm"} --set KMIR_HASKELL_DIR ${(toString prev.mir-semantics) + "/lib/haskell"} --prefix PATH : ${prev.lib.makeBinPath [ k-framework.packages.${prev.system}.k ]}
```

```suggestion
              wrapProgram $out/bin/kmir --set KMIR_LLVM_DIR "$out/lib/llvm" --set KMIR_HASKELL_DIR "$out/lib/haskell" --prefix PATH : ${prev.lib.makeBinPath [ k-framework.packages.${prev.system}.k ]}
```

bad things usually happen if a derivation tries to refer to itself via anything other than the`$out` variable.

---

I couldn't get this to work previously, and I tried this change just now, it doesn't route correctly. I think it `$out` is the python kmir dir, and that is not where the defs are:
```
$ kmir parse src/tests/integration/test-data/handwritten-mir/execution/arithm-simple.mir 
Traceback (most recent call last):
  File "/nix/store/l3129p42lrgj63m97mspx9qx502x7ygj-python3.10-kmir-0.1.0/bin/..kmir-wrapped-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/l3129p42lrgj63m97mspx9qx502x7ygj-python3.10-kmir-0.1.0/lib/python3.10/site-packages/kmir/__main__.py", line 22, in main
    execute(**vars(args))
  File "/nix/store/l3129p42lrgj63m97mspx9qx502x7ygj-python3.10-kmir-0.1.0/lib/python3.10/site-packages/kmir/__main__.py", line 39, in exec_parse
    kmir = KMIR(definition_dir, None)
  File "/nix/store/l3129p42lrgj63m97mspx9qx502x7ygj-python3.10-kmir-0.1.0/lib/python3.10/site-packages/kmir/kmir.py", line 45, in __init__
    check_dir_path(llvm_dir)
  File "/nix/store/b0197fwlh6szlhv1qsh7n99baz8lvvgc-python3.10-pyk-0.1.401/lib/python3.10/site-packages/pyk/utils.py", line 420, in check_dir_path
    raise ValueError(f'Directory does not exist: {path}')
ValueError: Directory does not exist: /nix/store/l3129p42lrgj63m97mspx9qx502x7ygj-python3.10-kmir-0.1.0/lib/llvm
```

---

Unless I am mistaken, we want this result:
`export KMIR_LLVM_DIR='/nix/store/3rbbjf78r362vam1gykb3nw37h754nz0-mir-semantics-a386aad115d93c92d2b1418c33086833ef17a098/lib/llvm'`
Not this result:
`export KMIR_LLVM_DIR='/nix/store/l3129p42lrgj63m97mspx9qx502x7ygj-python3.10-kmir-0.1.0/lib/llvm'`

Is there someway I can achieve that without referencing `prev.mir-semantics`? I couldn't get it to work.